eng/validation/build-arcadewithrepo.ps1

```diff
@@ -0,0 +1,265 @@
+Param(
+  [Parameter(Mandatory=$true)]
+  [string] $azdoOrg, 
+
+  [Parameter(Mandatory=$true)]
+  [string] $azdoProject,
+
+  [Parameter(Mandatory=$true)]
+  [int] $buildDefinitionId,
+
+  [Parameter(Mandatory=$true)]
+  [string] $azdoToken,
+
+  [Parameter(Mandatory=$true)]
+  [string] $githubUser,
+  
+  [Parameter(Mandatory=$true)]
+  [string] $githubPAT,
+
+  [Parameter(Mandatory=$true)]
+  [string] $githubOrg, 
+
+  [Parameter(Mandatory=$true)]
+  [string] $targetRepoName,
+
+  [Parameter(Mandatory=$true)]
+  [string] $barToken, 
+
+  [string] $buildParameters = ''
+)
+
+set-strictmode -version 2.0
+$ErrorActionPreference = 'Stop'
+
+. $PSScriptRoot\..\common\tools.ps1
+. $PSScriptRoot\..\common\darc-init.ps1
+
+$arcadeSdkPackageName = 'Microsoft.DotNet.Arcade.Sdk'
+$arcadeSdkVersion = $GlobalJson.'msbuild-sdks'.$arcadeSdkPackageName
+
+# Get a temporary directory for a test root. Use the agent work folder if running under azdo, use the temp path if not.
+$testRootBase = if ($env:AGENT_WORKFOLDER) { $env:AGENT_WORKFOLDER } else { $([System.IO.Path]::GetTempPath()) }
+$testRoot = Join-Path -Path $testRootBase -ChildPath $([System.IO.Path]::GetRandomFileName())
+New-Item -Path $testRoot -ItemType Directory | Out-Null
+
+$days = 3
+$minTime = (Get-Date).AddDays(-$days)
+$buildReasonsList = @("batchedCI", "individualCI")
+
+function Get-LastKnownGoodBuildSha(
+    [string] $azdoOrg, 
+    [string] $azdoProject,
+    [int] $buildDefinitionId,
+    [string] $azdoToken)
+{
+
+    ## Have there been any builds in the last $days days? 
+    $base64AuthInfo = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(":${azdoToken}"))
+    $headers = @{"Authorization"="Basic $base64AuthInfo"}
+    $count = 0
+
+    foreach($reason in $buildReasonsList)
+    {
+        $uri = "https://dev.azure.com/${azdoOrg}/${azdoProject}/_apis/build/builds?api-version=5.1&statusFilter=completed&definitions=${buildDefinitionId}&reasonFilter=${reason}&minTime=${minTime}&`$top=1"
+        $reponse = (Invoke-WebRequest -Uri $uri -Headers $headers -Method Get) | ConvertFrom-Json
+        $count += $reponse.count
+    }
+
+    ## No? Then get the last known good build regardless of age
+    if($count -eq 0)
+    {
+        $contentArray = @()
+
+        foreach($reason in $buildReasonsList)
+        {
+            $uri = "https://dev.azure.com/${azdoOrg}/${azdoProject}/_apis/build/builds?api-version=5.1&resultsFilter=succeeded&definitions=${buildDefinitionId}&reasonFilter=${reason}&buildQueryOrder=finishTimeAscending&`$top=1"
```

Comment:
This can probably be extracted into a function as it's repeated below just with slightly different params.

Reply:
I extracted most of the uri into it's own function, and almost started to do the same with the query string parameters, but the function was going to have to take in so many parameters that it would've defeated the purpose of creating a method to return a string. Hope that's good enough. :/

Reply:
What I meant was that you could extract the whole call to the API to a function. 

https://github.com/dotnet/arcade-validation/pull/1360/files#diff-0c4c1d2780faad03a492cdb1cab92517R52-R60 and https://github.com/dotnet/arcade-validation/pull/1360/files#diff-0c4c1d2780faad03a492cdb1cab92517R71-R79 are duplicate. You just change the params to the build API and what you do with the resulting array.

Reply:
That would also allow us to add retries, etc in a single place if we see flakiness.

Reply:
Sorry Ricardo, I somehow completely missed these comments. I can combine those two, but there's a third API call in the same method with very similar (but different) parameters: https://github.com/dotnet/arcade-validation/pull/1360/files#diff-0c4c1d2780faad03a492cdb1cab92517R42

Are you okay with that one being left out of a method that wraps those two API calls into a method? I was trying to figure out how to get all three of them into a method and since they were taking different parameters, the method call was going to be as long as the query string parameter (probably longer, because this is Powershell, of course) :)

Reply:
Yeah, I'm fine with that.